### PR TITLE
[oneMKL][Sparse BLAS] Add preconditioned conjugate gradient in USM with two examples

### PR DESCRIPTION
**Pull request by spencerpatty**
# Existing Sample Changes
## Description

Switch from sycl::buffer to USM for PCG example, using ||r|| instead of dot(w,w) for convergence (which is what the previous sycl::buffer version used). Two implementations of PCG are provided:

1. using alpha/beta coefficients assembled host side -- uses BLAS and Sparse BLAS routines, but there are several explicit sync locations to assemble the coefficients before sendingthem into the BLAS routines.

2. using alpha/beta coefficients assembled on-the-fly with device side arrays -- uses both BLAS and Sparse BLAS routines, but two new APIs:  `axpy3()` and `axpby2()` to handle the input scalar arguments to be assembled on-the-fly:

```
//
// AXPBY2:
//
// Calculate y = alpha * x + beta1 / beta2 * y
//
// where beta1 and beta2 are device side arrays with scalar value in 0th element
//
template <typename dataType, typename intType>
sycl::event axpby2(sycl::queue q,
                   const intType n,
                   const dataType alpha, // host side scalar
                   const dataType *x,
                   const dataType *beta1_d, // device side scalar numerator
                   const dataType *beta2_d, // device side scalar denominator
                         dataType *y,
                   const std::vector<sycl::event> &deps = {})
{
    return q.submit([&](sycl::handler &cgh) {
        cgh.depends_on(deps);
        auto kernel = [=](sycl::item<1> item) {
            const int row = item.get_id(0);

            const dataType beta = beta1_d[0] / beta2_d[0];
            y[row] = alpha * x[row] + beta * y[row];
        };
        cgh.parallel_for<class axpby2Class<dataType, intType>>(sycl::range<1>(n), kernel);
    });
}
```
and
```
//
// AXPY3:
//
// Calculate y = scale * alpha1 / alpha2 * x + y
//
// where alpha1 and alpha2 may be device side arrays with scalar value in 0th element
//
template <typename dataType,typename intType>
sycl::event axpy3(sycl::queue q,
                  const intType n,
                  const dataType scale,     // host side scalar 
                  const dataType *alpha1_d, // device side scalar numerator
                  const dataType *alpha2_d, // device side scalar denominator
                  const dataType *x,
                        dataType *y,
                  const std::vector<sycl::event> &deps = {})
{
    return q.submit([&](sycl::handler &cgh) {
        cgh.depends_on(deps);
        auto kernel = [=](sycl::item<1> item) {
            const int row = item.get_id(0);
            const dataType alpha = scale * alpha1_d[0] / alpha2_d[0];
            y[row] = alpha * x[row] + y[row];
        };
        cgh.parallel_for<class axpy3Class<dataType, intType>>(sycl::range<1>(n), kernel);
    });
}
```

Validation is added to the examples so that it returns success or not.  The matrices are modified to be diagonally dominant, so that the PCG process converges quickly.  

## External Dependencies

None

## Type of change

Please delete options that are not relevant. Add a 'X' to the one that is applicable. 

- [X] New feature (non-breaking change which adds functionality)

Also resolves request to add validation to example stored in internal to Intel JIRA [MKLD-18015](https://jira.devtools.intel.com/browse/MKLD-18015).

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] Command Line
- [ ] oneapi-cli
- [ ] Visual Studio
- [ ] Eclipse IDE
- [ ] VSCode
- [X] When compiling the compiler flag "-Wall -Wformat-security -Werror=format-security" was used

### CODE REVIEW

File: Libraries/oneMKL/sparse_conjugate_gradient/README.md [hunk 1]
```diff
@@ -17,7 +17,7 @@ Sparse Conjugate Gradient uses oneMKL sparse linear algebra routines to solve a
...
-oneMKL sparse routines use a two-stage method where the sparse matrix is analyzed to prepare subsequent calculations (the _optimize_ step). Sparse matrix-vector multiplication and triangular solves (`gemv` and `trsv`) are used to implement the main loop, along with vector routines from BLAS.
+oneMKL sparse routines use a two-stage method where the sparse matrix is analyzed to prepare subsequent calculations (the _optimize_ step). Sparse matrix-vector multiplication and triangular solves (`gemv` and `trsv`) are used to implement the main loop, along with vector routines from BLAS. Two implementations are provided: The first implementation, in `sparse_cg.cpp`, has several places where a device to host copy and wait are initiated to allow the alpha and beta coefficients to be initiated in the BLAS vector routines as host scalars.  The second implementation, in `sparse_cg2.cpp`, keeps the coefficients for alpha and beta on the device, which require that custom axpby2 and axpy3 fucntions are written to handle the construction of alpha and beta coefficients on-the-fly from the device. This removes some of the synchronization points that are seen in the first implementation.
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/README.md:20]:
```suggestion
oneMKL sparse routines use a two-stage method where the sparse matrix is analyzed to prepare subsequent calculations (the _optimize_ step). Sparse matrix-vector multiplication and triangular solves (`gemv` and `trsv`) are used to implement the main loop, along with vector routines from BLAS. Two implementations are provided: The first implementation, in `sparse_cg.cpp`, has several places where a device to host copy and wait are initiated to allow the alpha and beta coefficients to be initiated in the BLAS vector routines as host scalars.  The second implementation, in `sparse_cg2.cpp`, keeps the coefficients for alpha and beta on the device, which require that custom axpby2 and axpy3 functions are written to handle the construction of alpha and beta coefficients on-the-fly from the device. This removes some of the synchronization points that are seen in the first implementation.
```

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp [hunk 2]
```diff
@@ -112,181 +289,264 @@ void run_sparse_cg_example(const sycl::device &dev)
...
+
+    // create arrays for help
+    intType *ia_d    = sycl::malloc_device<intType>(n+1, q);
+    intType *ja_d    = sycl::malloc_device<intType>(nnz, q);
+    dataType *a_d    = sycl::malloc_device<dataType>(nnz, q);
+    dataType *x_d    = sycl::malloc_device<dataType>(n, q);
+    dataType *b_d    = sycl::malloc_device<dataType>(n, q);
+    dataType *r_d    = sycl::malloc_device<dataType>(n, q); // r = b - A*x
+    dataType *z_d    = sycl::malloc_device<dataType>(n, q); // z = M^{-1} * r
+    dataType *p_d    = sycl::malloc_device<dataType>(n, q); // p = ?? z
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp:343]:
```suggestion
    dataType *p_d    = sycl::malloc_device<dataType>(n, q); // p = z + beta * p
```

Comment (noffermans) [reply Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp:343]: Or just "search direction"?

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp [hunk 3]
```diff
@@ -112,181 +289,264 @@ void run_sparse_cg_example(const sycl::device &dev)
...
-    sycl::buffer<fp, 1> y_buffer(nrows);
-    sycl::buffer<fp, 1> d_buffer(nrows);
-    sycl::buffer<fp, 1> temp_buffer(1);
+    // device side aliases scattered by width elements each
+    dataType *normr_h  = temp_h;
+    dataType *normr_d  = temp_d;
+    dataType *rtz_h    = temp_h+1*width;
+    dataType *rtz_d    = temp_d+1*width;
+    //dataType *oldrtz_h = temp_h+2*width;
+    //dataType *oldrtz_d = temp_d+2*width;
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp:362]:
```suggestion
```

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp [hunk 4]
```diff
@@ -112,181 +289,264 @@ void run_sparse_cg_example(const sycl::device &dev)
...
+    dataType *r_d    = sycl::malloc_device<dataType>(n, q); // r = b - A*x
+    dataType *z_d    = sycl::malloc_device<dataType>(n, q); // z = M^{-1} * r
+    dataType *p_d    = sycl::malloc_device<dataType>(n, q); // p = ?? z
+    dataType *t_d    = sycl::malloc_device<dataType>(n, q); // helper array
+    dataType *d_d    = sycl::malloc_device<dataType>(n, q); // diag(A)
+    dataType *invd_d = sycl::malloc_device<dataType>(n, q); // inv(diag(A))
+
+    const intType width = 8; // width * sizeof(dataType) >= cacheline size (64 Bytes)
+    dataType *temp_d = sycl::malloc_device<dataType>(4*width, q);
+    dataType *temp_h = sycl::malloc_host<dataType>(4*width, q);
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp:350]:
```suggestion
    dataType *temp_d = sycl::malloc_device<dataType>(3*width, q);
    dataType *temp_h = sycl::malloc_host<dataType>(3*width, q);
```
If H2D transfer of  `oldrtz` between is not needed.

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp [hunk 5]
```diff
@@ -112,181 +289,264 @@ void run_sparse_cg_example(const sycl::device &dev)
...
-    sycl::buffer<fp, 1> temp_buffer(1);
+    // device side aliases scattered by width elements each
+    dataType *normr_h  = temp_h;
+    dataType *normr_d  = temp_d;
+    dataType *rtz_h    = temp_h+1*width;
+    dataType *rtz_d    = temp_d+1*width;
+    //dataType *oldrtz_h = temp_h+2*width;
+    //dataType *oldrtz_d = temp_d+2*width;
+    dataType *pAp_h    = temp_h+3*width;
+    dataType *pAp_d    = temp_d+3*width;
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp:364]:
```suggestion
    dataType *pAp_h    = temp_h+2*width;
    dataType *pAp_d    = temp_d+2*width;
```

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp [hunk 6]
```diff
@@ -0,0 +1,646 @@
...
+
+    // device side aliases scattered by width elements each
+    dataType *normr_h  = temp_h;
+    dataType *normr_d  = temp_d;
+    //dataType *rtz_h    = temp_h+1*width;
+    dataType *rtz_d    = temp_d+1*width;
+    //dataType *oldrtz_h = temp_h+2*width;
+    dataType *oldrtz_d = temp_d+2*width;
+    //dataType *pAp_h    = temp_h+3*width;
+    dataType *pAp_d    = temp_d+3*width;
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp:438]: Do you want to adjust the alloc size on the host and delete the commented lines?
Or is that to show how we skip some syncing points when using device size scalars?

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp [hunk 7]
```diff
@@ -0,0 +1,646 @@
...
+
+    // create arrays for help
+    intType *ia_d    = sycl::malloc_device<intType>(n+1, q);
+    intType *ja_d    = sycl::malloc_device<intType>(nnz, q);
+    dataType *a_d    = sycl::malloc_device<dataType>(nnz, q);
+    dataType *x_d    = sycl::malloc_device<dataType>(n, q);
+    dataType *b_d    = sycl::malloc_device<dataType>(n, q);
+    dataType *r_d    = sycl::malloc_device<dataType>(n, q); // r = b - A*x
+    dataType *z_d    = sycl::malloc_device<dataType>(n, q); // z = M^{-1} * r
+    dataType *p_d    = sycl::malloc_device<dataType>(n, q); // p = ?? z
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp:417]:
```suggestion
    dataType *p_d    = sycl::malloc_device<dataType>(n, q); // p = z + beta * p
```
Or "search direction"?

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp (hunk 2 above)
Comment (spencerpatty) [reply Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp:343]: search direction is preferrable

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp [hunk 8]
```diff
@@ -112,181 +289,263 @@ void run_sparse_cg_example(const sycl::device &dev)
...
+    const intType nnz = ia_h[n]; // assumes zero indexing
+
+    // create arrays for help
+    intType *ia_d    = sycl::malloc_device<intType>(n+1, q);  // matrix rowptr
+    intType *ja_d    = sycl::malloc_device<intType>(nnz, q);  // matrix columns
+    dataType *a_d    = sycl::malloc_device<dataType>(nnz, q); // matrix values
+    dataType *x_d    = sycl::malloc_device<dataType>(n, q);   // solution
+    dataType *b_d    = sycl::malloc_device<dataType>(n, q);   // right hand side
+    dataType *r_d    = sycl::malloc_device<dataType>(n, q);   // residual
+    dataType *z_d    = sycl::malloc_device<dataType>(n, q);   // preconditioned residualr
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp:342]:
```suggestion
    dataType *z_d    = sycl::malloc_device<dataType>(n, q);   // preconditioned residual
```

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp [hunk 9]
```diff
@@ -0,0 +1,644 @@
...
+    const intType nnz = ia_h[n]; // assumes zero indexing
+
+    // create arrays for help
+    intType *ia_d    = sycl::malloc_device<intType>(n+1, q);  // matrix rowptr
+    intType *ja_d    = sycl::malloc_device<intType>(nnz, q);  // matrix columns
+    dataType *a_d    = sycl::malloc_device<dataType>(nnz, q); // matrix values
+    dataType *x_d    = sycl::malloc_device<dataType>(n, q);   // solution
+    dataType *b_d    = sycl::malloc_device<dataType>(n, q);   // right hand side
+    dataType *r_d    = sycl::malloc_device<dataType>(n, q);   // residual
+    dataType *z_d    = sycl::malloc_device<dataType>(n, q);   // preconditioned residualr
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp:416]:
```suggestion
    dataType *z_d    = sycl::malloc_device<dataType>(n, q);   // preconditioned residual
```

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp [hunk 10]
```diff
@@ -0,0 +1,644 @@
...
+        oneapi::mkl::sparse::release_matrix_handle(q, &A).wait();
+        return 1;
+    }
+
+    q.wait();
+
+    //  clean up USM memory allocations
+    sycl::free(ia_h, q);
+    sycl::free(ja_h, q);
+    sycl::free(a_h, q);
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp:585]:
```suggestion
    sycl::free(a_h, q);
    sycl::free(x_h, q);
    sycl::free(b_h, q);
```

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp [hunk 11]
```diff
@@ -0,0 +1,644 @@
...
+    sycl::free(ia_d, q);
+    sycl::free(ja_d, q);
+    sycl::free(a_d, q);
+    sycl::free(x_d, q);
+    sycl::free(b_d, q);
+    sycl::free(t_d, q);
+    sycl::free(d_d, q);
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp:592]:
```suggestion
    sycl::free(d_d, q);
    sycl::free(invd_d, q);
```

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp [hunk 12]
```diff
@@ -0,0 +1,644 @@
...
+
+    //  clean up USM memory allocations
+    sycl::free(ia_h, q);
+    sycl::free(ja_h, q);
+    sycl::free(a_h, q);
+    sycl::free(ia_d, q);
+    sycl::free(ja_d, q);
+    sycl::free(a_d, q);
+    sycl::free(x_d, q);
+    sycl::free(b_d, q);
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp:590]:
```suggestion
    sycl::free(b_d, q);
    sycl::free(r_d, q);
    sycl::free(z_d, q);
    sycl::free(p_d, q);
```

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp [hunk 13]
```diff
@@ -112,181 +289,263 @@ void run_sparse_cg_example(const sycl::device &dev)
...
     }
 
-    mkl::sparse::release_matrix_handle(main_queue, &handle);
-    main_queue.wait();
+    q.wait();
+
+    //  clean up USM memory allocations
+    sycl::free(ia_h, q);
+    sycl::free(ja_h, q);
+    sycl::free(a_h, q);
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp:523]:
```suggestion
    sycl::free(a_h, q);
    sycl::free(x_h, q);
    sycl::free(b_h, q);
```

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp [hunk 14]
```diff
@@ -112,181 +289,263 @@ void run_sparse_cg_example(const sycl::device &dev)
...
+    sycl::free(ia_d, q);
+    sycl::free(ja_d, q);
+    sycl::free(a_d, q);
+    sycl::free(x_d, q);
+    sycl::free(b_d, q);
+    sycl::free(t_d, q);
+    sycl::free(d_d, q);
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp:530]:
```suggestion
    sycl::free(d_d, q);
    sycl::free(invd_d, q);
```

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp [hunk 15]
```diff
@@ -0,0 +1,644 @@
+//==============================================================
+// Copyright © 2020 Intel Corporation
```
Comment (noffermans) [inline Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg2.cpp:2]:
```suggestion
// Copyright © 2024 Intel Corporation
```

**Review by noffermans — APPROVED** [review]
There might be some `sylc::free` missing that I commented about.
Otherwise, these examples look great! Approved.

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp (hunk 14 above)
Comment (spencerpatty) [reply Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp:530]: good catch

File: Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp [hunk 16]
```diff
@@ -112,181 +289,263 @@ void run_sparse_cg_example(const sycl::device &dev)
...
+
+    //  clean up USM memory allocations
+    sycl::free(ia_h, q);
+    sycl::free(ja_h, q);
+    sycl::free(a_h, q);
+    sycl::free(ia_d, q);
+    sycl::free(ja_d, q);
+    sycl::free(a_d, q);
+    sycl::free(x_d, q);
+    sycl::free(b_d, q);
```
[the remark this answers is not in the document]
Comment (spencerpatty) [reply Libraries/oneMKL/sparse_conjugate_gradient/sparse_cg.cpp:528]: oh my, I missed a ton of these

**Comment by spencerpatty** [comment]
All reviewed issues have been resolved, the example runs great locally.  In my opinion, it is ready for merge.

**Pull request merged**